lib/pbench/server/api/resources/__init__.py

```diff
@@ -0,0 +1,694 @@
+import json
+from datetime import datetime
+from enum import Enum
+from http import HTTPStatus
+from logging import Logger
+from typing import Any, AnyStr, Callable, Dict, List, Union
+from flask.wrappers import Request, Response
+
+from dateutil import parser as date_parser
+from flask import request
+from flask_restful import Resource, abort
+from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
+from pbench.server.database.models.datasets import Dataset
+from pbench.server.database.models.users import User
+
+# A type defined to conform to the semantic definition of a JSON structure
+# with Python syntax.
+JSONSTRING = str
+JSONNUMBER = Union[int, float]
+JSONVALUE = Union["JSONOBJECT", "JSONARRAY", JSONSTRING, JSONNUMBER, bool, None]
+JSONARRAY = List[JSONVALUE]
+JSONOBJECT = Dict[JSONSTRING, JSONVALUE]
+JSON = JSONVALUE
+
+
+class UnauthorizedAccess(Exception):
+    """
+    The user is not authorized for the requested operation on the specified
+    resource.
+    """
+
+    def __init__(self, user: str, operation: "API_OPERATION", owner: str, access: str):
+        self.user = user
+        self.operation = operation
+        self.owner = owner
+        self.access = access
+
+    def __str__(self) -> str:
+        return f"User {self.user} is not authorized to {self.operation} a resource owned by {self.owner} with {self.access} access"
```

Comment:
I prefer exceptions with machine readable parameters rather than just having a string representation.

I'll think about adding some generic text to the base exception classes; although probably not here since the idea was to keep this as narrowly focused as feasible on the refactoring.

Reply:
> I prefer exceptions with machine readable parameters

I suspected that that might be the case.

However, you can still set the stringization at init-time and leave the handling of it up to the base class.  Of course, if someone changes the attributes "after the fact", the string won't be updated, but I think that's acceptable in light of the simplification of the code.  (Ideally, the attributes would be read-only after initialization or they would have setters which updated the string as well, but that's a crazy amount of superstructure relative to what we actually need.)

Reply:
_[Unresolving, because GitHub won't let me start a _new_ conversation for the same code range....]_

Are you satisfied with how this message will look if either `user` or `owner` are `None`?

Reply:
I'm trying really hard not to tackle that additional level of cleanup in this PR.